### ISSUE TITLE
Discussion on new organization of shr_stream_mod and related code

### ISSUE THREAD

**Issue by billsacks** [ESMCI/cime#1802]
In #1767 , @mvertens moved shr_stream_mod and related code out of `src/share/util` into `src/components/data_comps/dshare`. There was some discussion of this move in the comments of #1767 (https://github.com/ESMCI/cime/pull/1767#issuecomment-318493072 and then later comments by @mvertens @jedwards4b and @rljacob ), but I'm not sure that my points were understood correctly. Since that PR has been merged, I wanted to continue the discussion in a new issue. To be clear: I am perfectly happy to be overruled on this point, but I just want to make sure that my points come across clearly before this decision is finalized.

Note that @mvertens set the build up so that `src/components/data_comps/dshare` is compiled into the share library, similarly to what is done for `src/drivers/mct/shr`. 

I have two related concerns with the new organization:

1. For my first few years of working with CESM, I found it confusing that the share library was built from code from two completely different places: the share directory itself as well as the shr subdirectory of the drv code. This move introduces a third location included in the share library. It would be easier to understand how the source code is built into libraries – and thus, crucially, what modules are available to components using cime – if the organization into libraries more closely matched the source code directory structure.

2. For someone not intimately familiar with the cime source code, a logical place to look for modules that can be utilized by component models is `src/share`. This will now fail to turn up the stream-related code - which can be a very useful set of utilities for active components as well as data components. Putting this code under `src/components/data_comps/dshare` sends the message that this code is only meant for use by data models, which is not at all the case. I am concerned that the new organization will therefore cause component model developers to overlook this code, or at the very least have a hard time finding it.

Here are the uses of the 4 modules included in this new `dshare` directory, in CESM:

• shr_strdata_mod: This is the key one: it is used extensively by CLM, as well as by POP and CICE:

  ```
  ./cime/src/components/data_comps/datm/datm_comp_mod.F90
  ./cime/src/components/data_comps/desp/desp_comp_mod.F90
  ./cime/src/components/data_comps/dice/dice_comp_mod.F90
  ./cime/src/components/data_comps/dlnd/dlnd_comp_mod.F90
  ./cime/src/components/data_comps/docn/docn_comp_mod.F90
  ./cime/src/components/data_comps/drof/drof_comp_mod.F90
  ./cime/src/components/data_comps/dwav/dwav_comp_mod.F90
  ./cime/src/share/util/shr_scam_mod.F90
  ./components/cice/src/drivers/cesm/ice_prescribed_mod.F90
  ./components/clm/src/biogeochem/ch4FInundatedStreamType.F90
  ./components/clm/src/biogeochem/CNFireBaseMod.F90
  ./components/clm/src/biogeochem/SatellitePhenologyMod.F90
  ./components/clm/src/biogeophys/UrbanTimeVarType.F90
  ./components/clm/src/main/ndepStreamMod.F90
  ./components/clm/src_clm40/main/ndepStreamMod.F90
  ./components/pop/source/strdata_interface_mod.F90
  ```

• shr_tInterp_mod: only used by shr_strdata_mod.F90 (so this should go in whatever directory shr_strdata_mod.F90 goes in)

• shr_stream_mod: In addition to shr_dmodel_mod.F90 and shr_strdata_mod.F90, this is used directly by two modules in CLM

• shr_dmodel_mod: in addition to the data models, this is used by:
  - cime/src/share/util/shr_scam_mod.F90
  - cime/src/share/util/shr_strdata_mod.F90
    - This use suggests to me that some reorganization of these modules may be needed: active components use shr_strdata_mod, which uses shr_dmodel_mod, which doesn't make sense conceptually
  - components/cice/src/drivers/cesm/ice_prescribed_mod.F90

**Comment by ekluzek** [comment]
Just to confirm Bill's comment "Putting this code under src/components/data_comps/dshare sends the message that this code is only meant for use by data models, which is not at all the case.". CLM uses stream code for several things internal to it (lightning, ndep, pop-density, finundation, urban heating). I imagine other active components might or could (and maybe should) use it. The streams code is good to take advantage of because of it's full functionality and it's better to use it than custom code for a particular component. I've been advocating that we use steams in CLM for things that make sense.

**Comment by apcraig** [comment]
I agree with Bill's sentiment.  And I think there should be a well defined plan for the directory structure.  It seems the directory structures are quite ad-hoc and still constantly changing, in big and small ways.  I had advocated for a long time that there should be a broad and high level plan for the design.  Process leveraging primarily "incremental" and "fix later" feels like it moves quickly but it actually has a large cost.  If CIME is going to be a community tool, there really needs to be a clean, consistent, and documented design and implementation across the board of many things (directory structure, compsets, components, drivers, share code, build, run, overall scripts, etc, etc) with a the ability to extend taken into account.  

Sorry, that was a bit off-issue.  Specifically addressing the location of the stream code, I agree that should not be under data_comps.

**Comment by jedwards4b** [comment]
@agsalin Can you tell us where you are with the cmake project and how it might affect directory structure?

**Comment by billsacks** [comment]
@mvertens I think there was agreement to move the contents of cime/src/components/data_comps/dshare/ back into cime/src/share. I can't remember if there was agreement on whether these modules should go back into cime/src/share/util, some new subdirectory of share, or some new subdirectory of share/util.

Were you still planning to do this move back at some point?

**Comment by billsacks** [comment]
When this is done, the paths referenced in `cime/doc/source/data_models/design-details.rst` should be updated accordingly.

**Comment by mvertens** [comment]
Frankly, I'm still not comfortable moving things back. I've been working a
lot with the data models recently and its extremely nice to have the share
code of the data models be located with the data models. I'd like to open
up this discussion again.

**Comment by billsacks** [comment]
@mvertens My feelings on this haven't changed. I see the point that the data models are the biggest users of this shared code, but:

1. In terms of the number of people looking for the code, I'd guess that there are many more scientist-developers of prognostic components who want to find this code - since I'd suspect that few people outside of CSEG muck with the data model code. And I'd argue that code organization is more important for outside developers than it is for CSEG members (CSEG members will be able to find the code anywhere, but outside developers may not have a clue as to where to find it).

2. Even if it weren't for point (1), I'd argue that code shared between multiple components belongs in a directory structure that indicates it as being shared code - not in a directory structure tied to the component(s) that happen to use it most heavily.

**Comment by billsacks** [comment]
However, I'll reiterate: I am perfectly happy to be overruled on this point if my views are now in the minority.

**Comment by apcraig** [comment]
I agree 100% with Bill's points.  A very practical situation where this becomes highly problematic is where a non-data model is using the data model share code, but maybe the data models are not available.  Maybe the model has been pulled into a different coupled system, or maybe there is a standalone mode where the prognostic model is used outside cime, or maybe there is a model where we cime is used but data models are not part of the directory structure because they are not needed.  This will cause all kinds of headaches.  Separately, it seems to me keeping multi-model share code within the data model directory is just fundamentally wrong.  It's like a subterranean cable with regard to source code management.

**Comment by mvertens** [comment]
But there is share code in the driver directory - that only pertains to
driver code. If a prognostic component wants to use share code - then it
seems to me that they should be able to simply point to the share code. I
realize that I'm in the minority here and I'm happy to go with the majority
viewpoint ( I just have not had a chance to move the directory yet)  - but
could more people please weigh in? @rjacob, @gold2718, @jfouca, @jedwards4b
- what are your thoughts?

**Comment by gold2718** [comment]
TL;DR: I vote that the dshare code stay where it is but be built as a separate library and that mct/shr be cleaned up but not built as part of the share library.

> Maybe the model has been pulled into a different coupled system, or maybe there is a standalone mode where the prognostic model is used outside cime, or maybe there is a model where we cime is used but data models are not part of the directory structure because they are not needed. 

I do not understand this bit. It seems to me that if you want to use CIME code outside of CIME, it is your responsibility to move that code with your model. If that is a true statement, I do not see how it matters from where in CIME you pull your code. What am I missing?

Personally, I would put a higher priority on keeping the code in a logical structure which in other SE organizations I have been a part of means the lowest root of the source tree which covers all usage (i.e., packages do not install their header files in /usr/include unless necessary).
IMHO, modifying the structure to facilitate dissecting CIME seems like a pretty low priority.

Now, having said all that, I agree with @billsacks that it is confusing if the library structure does not approximately match the directory structure. Why not build a dshare library for use by the data models?

The case of the the driver/mct/shr code is more complicated (again, IMHO). Some of this code is really MCT interface code (e.g., seq_cdata_setptrs) and should be located in a more appropriately named library and built as part of MCT. Other code (e.g., seq_io_read_mod) is only used internally by MCT and should not be in a shr directory. And, finally, there is code which is used by components but _not_ by MCT (e.g., seq_drydep_mod). This code should be in src/share, not in MCT.

**Comment by billsacks** [comment]
@gold2718 I'm having trouble reconciling your summary:

> I vote that the dshare code stay where it is but be built as a separate library

with

>  I would put a higher priority on keeping the code in a logical structure which in other SE organizations I have been a part of means the lowest root of the source tree which covers all usage 

Given that these modules are used by active components as well as data components, how would you justify having this code nested under "data_comps" given your "which covers all usage" piece?

**Comment by jedwards4b** [comment]
I think that I agree with Bill here, if the code is used by components outside of cime as well as cime components it should be in the share directory, perhaps in a different subdirectory and library.  One key point is that there is code in the share directory that uses this code which sets up a kind of circular dependency between the libraries.   I would also agree with @billsacks that although @gold2718 makes a good case, he seems to contradict himself with his preference.

**Comment by apcraig** [comment]
I also agree that the share code under the driver maybe should never have been there and that we should review the status of all the share code.  But, I also think there are some differences and that it might make some sense to keep it separate.  I view the generic share code as code that could be used by any model whether run standalone, in CESM, or otherwise.  The driver share code should really only contain share code that is needed when coupling within CESM/CIME.  Some code has ended up in the wrong place (share vs driver/share) under that definition, and I have pointed that out in the past at times.  And maybe some of the code in driver/share is more usable than just in CESM coupling.  It's worthwhile reviewing why the two directories exist and whether we feel that should continue.  But to me it's clear that the "data model" share code being used by prognostic models belongs in the generic share area as it could be used by more than data models and it may be needed in applications outside CESM.

**Comment by jedwards4b** [comment]
If someone is going to take on this restructure I would reference #852 as well.

**Comment by gold2718** [comment]
Sorry for contradicting myself (that message was written by my evil twin, Skippy), let me try again.

I think that shr_strdata_mod is the only dshare module used by active components. This means that per my formula, it should be moved to src/share (src/share/dshare?).

What I did not notice before is that shr_strdata_mod uses other dshare modules (through evil naked use statements). Therefore, the only way to build the share library with shr_strdata_mod in it is to also move the other dshare modules. So I change that part of my vote to moving src/components/data_comps/dshare to src/share/dshare.

I still think that mct/shr should be refactored for clarity.

**Comment by mvertens** [comment]
I will plan to issue a PR within the next few days with the move of
components/data_comps/dshare to share/dshare. Is there agreement that this
is the right place?

**Comment by billsacks** [comment]
@mvertens Thank you. I'm okay with share/dshare, though once this gets moved out of data_comps, it's less obvious what the 'd' stands for. So maybe something like share/data_share would be better?

@apcraig Good point about the shr stuff under the driver is only applicable when running with the mct-based driver/coupler.

**Comment by rljacob** [comment]
Catching up with this.  I agree with Bill and Tony: if its used by non-data models, it should be moved out of components/data_comps/.

Can you should put everything in src/share/util  instead of making src/share/dshare?

**Comment by apcraig** [comment]
Why don't we just drop the data model share code into share/util?  At this point, we don't really have a good basis for breaking the share code into different subdirectories.  If we are going to create subdirectories for the share code, I am fine with that, but I don't think it should be adhoc.  This feels adhoc.

**Comment by rljacob** [comment]
Heh.  Tony and I did not coordinate that.

**Comment by rljacob** [comment]
If we do this, would https://github.com/ESMCI/cime/issues/1713 become a problem again?

**Comment by mvertens** [comment]
Yes. This would be a problem again. This was not an adhoc decision. We had
talked about separating the data model share code at several cime meetings
because of #1713. My preference is to keep it parallel to utilize with a
data_share directory name.

**Comment by apcraig** [comment]
It seems like we could fix the circular dependence.  I wasn't aware of that issue.  How are things built now, I guess the sequence is share/util, driver/share, datm/share?  I actually thought all the share code was more or less built together at the same time and so it wouldn't matter where the code is actually located.  A circular dependence would be if seq_flds used shr_dmodel and I don't think that's the case.  Or is there really a circular dependency in the build?  It seems like #1713 is addressing a concern about extensibility away from mct coupling and the issue brought up there suggests that the seq_flds use in shr_dmodel has to be removed for that extensibility to be supported.  It should be possible for the shr_dmodel_readgrid interface to be modified so the gGrid field lists are passed in.  Or even for ggrid to be initialized with hardcoded field names.  If you look at the current implementation, we initialize ggrid with seq_flds field lists then we explicitly hardcode those names here,

     nlon  = mct_aVect_indexRA(gGridRoot%data,'lon')
     nlat  = mct_aVect_indexRA(gGridRoot%data,'lat')
...

To be consistent, we might as well hardcode the field names in the ggrid init in shr_dmodel_mod.  We have hardcoded them in that same routine anyway.  This actually brings up a whole host of other questions.  But in my opinion, there is no reason to keep seq_flds dependency in shr_dmodel.  We gain nothing by doing that right now.

**Comment by apcraig** [comment]
Let me just clarify that I think the data share code should NOT depend on the driver share code for extensibility and reuse.  We should fix that if possible.  If the seq_flds reference in shr_dmodel is the only issue, then that is easy to fix.  Separately, we can still discuss whether that data model share code should or should not be in it's own directory.

**Comment by mvertens** [comment]
@apcraig - I agree with your last point. I also think you are bringing up a good suggestion for removing the fixing the reference to seq_flds in shr_dmodel. I still think that the data model share code should be in its own directory since its not just generic utility code - but is specific to the data model functionality. Others please comment if you agree or disagree.

**Comment by jedwards4b** [comment]
I agree.

**Comment by rljacob** [comment]
Will the code in this new dir be built as a separate library?  Or together with the other share code?

The part that everyone uses is the stream capability right?   How about if we name the directory "src/share/streams" and this is another tool CIME offers which happens to be used by the data models and some other models.

**Comment by jedwards4b** [comment]
Seems like it should be a separate library or there isn't much point in putting it into a separate directory

**Comment by mvertens** [comment]
I agree that ultimately it should be a separate library - and this would be
a step to getting there. I think that separating data model functionality
helps clarify that this is more than just generic utility code. I'm happy
with Rob's calling it "streams".

**Comment by agsalin** [comment]
I like the idea of working towards **one library per directory, with no circular dependencies**. We don't currently have this, but it could be the guiding principle for new code and subsequent reorganizations. I think this in line with the consensus that was emerging already.
